src/extensions/prometheus/mod.rs

```diff
@@ -0,0 +1,90 @@
+mod rpc_metrics;
+
+use super::{Extension, ExtensionRegistry};
+use async_trait::async_trait;
+use serde::Deserialize;
+use std::iter;
+use std::net::SocketAddr;
+use substrate_prometheus_endpoint::init_prometheus;
+use substrate_prometheus_endpoint::Registry;
+use tokio::task::JoinHandle;
+
+use crate::utils::TypeRegistryRef;
+pub use rpc_metrics::RpcMetrics;
+
+pub async fn get_rpc_metrics(registry: &TypeRegistryRef) -> RpcMetrics {
+    let prometheus = registry.read().await.get::<Prometheus>();
+
+    match prometheus {
+        None => RpcMetrics::noop(),
+        Some(prom) => prom.rpc_metrics(),
+    }
+}
+
+pub struct Prometheus {
+    pub registry: Registry,
+    rpc_metrics: RpcMetrics,
+    pub exporter_task: JoinHandle<()>,
+}
+
+impl Drop for Prometheus {
+    fn drop(&mut self) {
+        self.exporter_task.abort();
+    }
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct PrometheusConfig {
+    pub port: u16,
+    pub listen_address: String,
+    pub prefix: Option<String>,
+    pub chain_label: Option<String>,
+}
+
+#[async_trait]
+impl Extension for Prometheus {
+    type Config = PrometheusConfig;
+
+    async fn from_config(config: &Self::Config, _registry: &ExtensionRegistry) -> Result<Self, anyhow::Error> {
+        Ok(Self::new(config.clone()))
+    }
+}
+
+impl Prometheus {
+    pub fn new(config: PrometheusConfig) -> Self {
+        let labels = config
+            .chain_label
+            .clone()
+            .map(|l| iter::once(("chain".to_string(), l.clone())).collect());
+        let prefix = match config.prefix {
+            Some(p) if p.is_empty() => None,
+            p => p,
+        };
+        let registry = Registry::new_custom(prefix, labels).expect("Can't happen");
```

Comment:
why it can't fail?

Reply:
It only can fail if the prefix is equal to `Some(p) where p.is_mepty()` and we take care of that in the match above

Reply:
can you explain it in the `expect`? otherwise the next reader is likely going to have the same question

Reply:
Sure, I think now it should be better